Error when joining string columns

```python
from datatable import *
X = Frame(A=['Ahoy yah matey!', 'hey'])
J = Frame(A=['hey'], B=['nope'])
J.key = 'A'
R = X[:, :, join(J)]
R.internal.check()
```
produces
```
AssertionError: Size of string data section in Column 1 does not correspond to the magnitude of the final offset: size = 33154, expected 400
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -110,6 +110,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 - A 0-row integer column can now be used as `i` in `DT[i, j]` (#1551).
 
+- A string column produced from a partial join now materializes correctly
+  (#1556).
+
 
 ### Changed
 
@@ -159,7 +162,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
   discovering and reporting bugs that were fixed in this release:
 
   [Pasha Stetsenko][] (#1316, #1443, #1481, #1542, #1551),
-  [Arno Candel][] (#1437, #1491, #1510, #1525, #1549),
+  [Arno Candel][] (#1437, #1491, #1510, #1525, #1549, #1556),
   [Michael Frasco][] (#1448),
   [Jonathan McKinney][] (#1451),
   [CarlosThinkBig][] (#1475),

--- a/c/column_string.cc
+++ b/c/column_string.cc
@@ -225,7 +225,7 @@ void StringColumn<T>::reify() {
   // If our rowindex is null, then we're already done
   if (ri.isabsent()) return;
   bool simple_slice = ri.isslice() && ri.slice_step() == 1;
-  bool ascending    = ri.isslice() && ri.slice_step() > 0;
+  bool ascending_slice = ri.isslice() && ri.slice_step() > 0;
 
   size_t new_mbuf_size = (ri.size() + 1) * sizeof(T);
   size_t new_strbuf_size = 0;
@@ -249,7 +249,7 @@ void StringColumn<T>::reify() {
       offs_dest[i] = data_src[i] - off0;
     }
 
-  } else if (ascending) {
+  } else if (ascending_slice) {
     // Special case: We can still do this in-place
     // (assuming the buffers are not read-only)
     if (!strbuf.is_writable())
@@ -289,6 +289,7 @@ void StringColumn<T>::reify() {
     T strs_size = 0;
     ri.iterate(0, nrows, 1,
       [&](size_t, size_t j) {
+        if (j == RowIndex::NA) return;
         strs_size += offs1[j] - offs0[j];
       });
     strs_size &= ~GETNA<T>();

--- a/tests/test_join.py
+++ b/tests/test_join.py
@@ -252,3 +252,14 @@ def test_join_view():
     assert res.shape == (2, 4)
     assert res.names == ("A", "B", "C", "BB")
     assert res.to_list() == [[1, 1], [3, 4], ['b', 'b'], [2, 2]]
+
+
+def test_issue1556():
+    X = dt.Frame(A=['Ahoy ye matey!', 'hey'])
+    J = dt.Frame(A=['hey'], B=['Avast'])
+    J.key = 'A'
+    R = X[:, :, join(J)]
+    R.internal.check()
+    assert R.shape == (2, 2)
+    assert R.to_dict() == {"A": ["Ahoy ye matey!", "hey"],
+                           "B": [None, "Avast"]}

